lib/synapse/service_watcher/zookeeper/zookeeper.rb

```diff
@@ -160,11 +177,19 @@ def watch
           @watcher = @zk.register(@discovery['path'], &watcher_callback)
         end
 
-        # Verify that we actually set up the watcher.
-        existed = with_retry(@retry_policy.merge({'retriable_errors' => ZK_RETRIABLE_ERRORS})) do |attempts|
-          log.info "synapse: zk exists at #{@discovery['path']} for #{attempts} times"
-          @zk.exists?(@discovery['path'], :watch => true)
+        # There are two cases when new watch needs to be created in zk, otherwise watch already exists.
+        # 1. Create watch in zk when zk event is nil
+        # 2. Create watch again when current event is triggered by existing watch. 
+        existed = true
+        need_exists_watch = zk_event.nil? or zk_event.node_changed? or zk_event.node_created? or zk_event.node_deleted?
```

Comment:
nit: use `||` instead of `or` for consistency

Reply:
will update